app/main.py

```diff
@@ -1,34 +1,164 @@
+        if self.start_row == self.end_row:
+            return [
+                (self.start_row, column)
+                for column in range(self.start_column, self.end_column + 1)
+            ]
+        else:
+            return [
+                (row, self.start_column)
+                for row in range(self.start_row, self.end_row + 1)
+            ]
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+
+    def fire(self, row: int, column: int) -> None:
+        deck = self.get_deck(row, column)
+        if deck:
+            deck.is_alive = False
+            if not any(deck.is_alive for deck in self.decks):
+                self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple]) -> None:
+        self.ship_instances = [Ship(start, end) for start, end in ships]
+        self.field = {}
+        for ship in self.ship_instances:
+            for deck in ship.decks:
+                if (deck.row, deck.column) in self.field:
+                    self.print_field()
+                    raise IncorectDataInput(
+                        f"{ship} can not be located as the cell"
+                        f" {deck.row, deck.column} is allready ocupied"
+                    )
+                self.field[(deck.row, deck.column)] = ship
+        self._validate_field()
+
+    def fire(self, location: tuple) -> str:
+        damaged_ship = self.field.get(location)
+        if damaged_ship and not damaged_ship.is_drowned:
+            damaged_ship.fire(*location)
+            return "Hit!" if not damaged_ship.is_drowned else "Sunk!"
+        return "Miss!"
+
+    def print_field(self) -> None:
+        field = {row: ["~" for i in range(10)] for row in range(10)}
+        for location, ship in self.field.items():
+            row, column = location
+            if ship.is_drowned:
+                field[row][column] = "x"
+            else:
+                if ship.get_deck(*location).is_alive:
+                    field[row][column] = "\u25A1"
+                else:
+                    field[row][column] = "*"
+
+        print("\n", end="")
+        print("\u25A1", end="   ")
+        for i in range(10):
```

Comment:
Do not use single letter variable names, use more descriptive ones instead of `i`.